cmd/lhsm-plugin-az-core/archive.go

```diff
@@ -17,101 +18,112 @@ type ArchiveOptions struct {
-	modTime := fileInfo.ModTime().Format("2006-01-02 15:04:05 -0700")
-
-	resp, err := blobURL.GetAccountInfo(context.Background())
-	if err != nil {
-		util.Log(pipeline.LogError, fmt.Sprintf("Archiving %s. Failed to account info: %s", o.BlobName, err.Error()))
-		return 0, err
-	}
 
-	hnsEnabledAccount := resp.Response().Header.Get("X-Ms-Is-Hns-Enabled") == "true"
-	if hnsEnabledAccount {
-		u, _ := url.Parse(fmt.Sprintf("https://%s.dfs.core.windows.net/%s/%s", o.AccountName, o.ContainerName, blobName))
-		dfsURL := azblob.NewBlockBlobURL(*u, p)
-		aclResp, err := dfsURL.GetAccessControl(ctx, nil, nil, nil, nil, nil, nil, nil, nil)
+	parents := strings.Split(o.BlobName, string(os.PathSeparator))
+	blobPath := o.MountRoot // Initialize the parent starting from root.
+	for _, currDir := range parents {
+		blobPath = path.Join(blobPath, currDir) //keep appending path to the url
+		blobURL := containerURL.NewBlockBlobURL(blobPath)
+		meta := azblob.Metadata{}
+
+		//Get owner, group and perms
+		fileInfo, err := os.Stat(blobPath)
 		if err != nil {
-			util.Log(pipeline.LogError, fmt.Sprintf("Archiving %s. Failed to get Access Control: %s", o.BlobName, err.Error()))
+			util.Log(pipeline.LogError, fmt.Sprintf("Archiving %s. Failed to get fileInfo: %s", o.BlobName, err.Error()))
 			return 0, err
 		}
-		acl = aclResp.XMsACL()
-	}
 
-	if !hnsEnabledAccount {
-		meta["Permissions"] = permissions
-		meta["ModTime"] = modTime
-		meta["Owner"] = owner
-		meta["Group"] = group
-	}
+		owner := fmt.Sprintf("%d", fileInfo.Sys().(*syscall.Stat_t).Uid)
+		permissions := fmt.Sprintf("%o", fileInfo.Mode())
+		group := fmt.Sprintf("%d", fileInfo.Sys().(*syscall.Stat_t).Gid)
+		modTime := fileInfo.ModTime().Format("2006-01-02 15:04:05 -0700")
+		var acl string
+
+		if o.HNSEnabled {
+			dfsEP, _ := url.Parse(fmt.Sprintf(dfsEndPoint+"%s%s", o.AccountName, o.ContainerName, o.ResourceSAS))
+			dfsURL := azblob.NewContainerURL(*dfsEP, p).NewBlockBlobURL(blobPath)
+			aclResp, err := dfsURL.GetAccessControl(ctx, nil, nil, nil, nil, nil, nil, nil, nil)
+			if stgErr, ok := err.(azblob.StorageError); err != nil || ok && stgErr.ServiceCode() != azblob.ServiceCodeBlobNotFound {
+				util.Log(pipeline.LogError, fmt.Sprintf("Archiving %s. Failed to get Access Control: %s", o.BlobName, err.Error()))
+				return 0, err
+			}
+			if err != nil {
+				acl = aclResp.XMsACL()
+			}
+		}
 
-	_, err = azblob.UploadFileToBlockBlob(
-		ctx,
-		file,
-		blobURL,
-		azblob.UploadToBlockBlobOptions{
-			BlockSize:   o.BlockSize,
-			Parallelism: o.Parallelism,
-			Metadata:    meta,
-		})
-
-	if err != nil {
-		util.Log(pipeline.LogError, fmt.Sprintf("Archiving %s. Failed to upload blob: %s", o.BlobName, err.Error()))
-		return 0, err
-	}
+		if !o.HNSEnabled {
+			meta["Permissions"] = permissions
+			meta["ModTime"] = modTime
+			meta["Owner"] = owner
+			meta["Group"] = group
+		}
+
+		if fileInfo.IsDir() {
+			meta["hdi_isfolder"] = "true"
+			_, err = azblob.UploadBufferToBlockBlob(
```

Comment:
For directories we can just use the put Blob API directly instead, since they have no body.

Reply:
yeah, I was trying to make file/dir look same, should have changed it earlier.